File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.js
let combi;
let set;
let banMap;

function solution(user_id, banned_id) {
    var answer = 0;
    
    banMap = new Map();
    combi = new Array(banned_id.length);
    set = new Set();
    
    for (let ban of banned_id) {
        banMap.set(ban, []);
        for (let user of user_id) {
            if (ban.length !== user.length) {
                continue;
            }
            let flag = true;
            for (let i = 0; i < ban.length; i++) {
                if (user[i] !== ban[i] && ban[i] !== "*") {
                    flag = false;
                    break;
                }
            }
            if (flag) {
                banMap.get(ban).push(user);
            }
        }
    }
    
    recursiveSolve(0, banned_id.length, banned_id);
    answer = set.size;
    
    return answer;
    
}

function recursiveSolve(idx, limit, banned_id) {
    
    if (idx === limit) {
        
        let tmpCombi = Array.from(combi);
        tmpCombi.sort();
        set.add(tmpCombi.join(""));
        
        return;
    }
    
    for (let ban of banMap.get(banned_id[idx])) {
        
        if (combi.includes(ban)) {
            continue;
        }
        
        combi[idx] = ban;
        recursiveSolve(idx + 1, limit, banned_id);
        combi[idx] = "";
    }
    
}

File: 프로그래머스/3/64064. 불량 사용자/불량 사용자.java
import java.util.*;
class Solution {
    
    private static String[] banned_id;
    private static String[] user_id;
    private static String[] combi;
    private static boolean[] visited;
    private static Set<String> result;
    
    public int solution(String[] user_id, String[] banned_id) {
        int answer = 0;
        
        this.banned_id = banned_id;
        this.user_id = user_id;
        combi = new String[banned_id.length];
        result = new HashSet<>();
        visited = new boolean[user_id.length];
        
        recursiveSolve(0);
        answer = result.size();
        
        return answer;
    }
    
    private static void recursiveSolve(int depth) {
        
        if (depth == banned_id.length) {
            int[] indices = new int[banned_id.length];
            for (int i = 0; i < banned_id.length; i++) {
                for (int j = 0; j < user_id.length; j++) {
                    if (combi[i].equals(user_id[j])) {
                        indices[i] = j;
                        break;
                    }
                }
            }
            Arrays.sort(indices);
            result.add(Arrays.toString(indices));
            return;
        }
        
        for (int i = 0; i < user_id.length; i++) {
            if (visited[i] || !isMatch(user_id[i], banned_id[depth])) {
                continue;
            }
            visited[i] = true;
            combi[depth] = user_id[i];
            recursiveSolve(depth + 1);
            visited[i] = false;
        }
        
    }
    
    private static boolean isMatch(String user, String banned) {
        if (user.length() != banned.length()) {
            return false;
        }
        for (int i = 0; i < user.length(); i++) {
            if (banned.charAt(i) != '*' && user.charAt(i) != banned.charAt(i)) {
                return false;
            }
        }
        return true;
    }
}